Modules/_uuidmodule.c

```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+
+    for (int i = 0; i < 4; i++) {
+        byte_to_hex(uuid->bytes[i], &str[i * 2]);
+    }
+    str[8] = '-';
+
+    for (int i = 4; i < 6; i++) {
+        byte_to_hex(uuid->bytes[i], &str[9 + (i - 4) * 2]);
+    }
+    str[13] = '-';
+
+    for (int i = 6; i < 8; i++) {
+        byte_to_hex(uuid->bytes[i], &str[14 + (i - 6) * 2]);
+    }
+    str[18] = '-';
+
+    for (int i = 8; i < 10; i++) {
+        byte_to_hex(uuid->bytes[i], &str[19 + (i - 8) * 2]);
+    }
+    str[23] = '-';
+
+    for (int i = 10; i < 16; i++) {
+        byte_to_hex(uuid->bytes[i], &str[24 + (i - 10) * 2]);
+    }
+
+    return PyUnicode_FromStringAndSize(str, 36);
+}
+
+static PyObject *
+Uuid_repr(PyObject *self)
+{
+    PyObject *str_obj = Uuid_str(self);
+    if (str_obj == NULL) {
+        return NULL;
+    }
+
+    // Get the class name (sadly can't use tp_name -- we don't need the full name)
+    PyObject *cls_name = PyObject_GetAttrString((PyObject *)Py_TYPE(self), "__name__");
+    if (cls_name == NULL) {
+        Py_DECREF(str_obj);
+        return NULL;
+    }
+
+    PyObject *repr = PyUnicode_FromFormat("%U('%U')", cls_name, str_obj);
+    Py_DECREF(str_obj);
+    Py_DECREF(cls_name);
+    return repr;
+}
+
+static int
+Uuid_setattr(PyObject *self, PyObject *name, PyObject *value)
+{
+    PyErr_SetString(PyExc_TypeError, "UUID objects are immutable");
+    return -1;
+}
+
+static PyObject *
+Uuid_get_urn(uuidobject *self, void *closure)
+{
+    PyObject *str_obj = Uuid_str((PyObject *)self);
+    if (str_obj == NULL) {
+        return NULL;
+    }
+
+    PyObject *urn = PyUnicode_FromFormat("urn:uuid:%U", str_obj);
+    Py_DECREF(str_obj);
+    return urn;
+}
+
+static Py_hash_t
+Uuid_hash(PyObject *self)
+{
+    uuidobject *uuid = (uuidobject *)self;
+    if (uuid->cached_hash != -1) {
+        // UUIDs are very often used in dicts/sets, so it makes
+        // sense to cache the computed hash (like we do for str)
+        return uuid->cached_hash;
+    }
+
+    Py_hash_t hash = Py_HashBuffer(uuid->bytes, 16);
+    uuid->cached_hash = hash;
+    return hash;
+}
+
+
+/*[clinic input]
+@classmethod
+_uuid.UUID._from_int
+
+    value: object
+    /
+
+Create a UUID from an integer value. Internal use only.
+[clinic start generated code]*/
+
+static PyObject *
+_uuid_UUID__from_int_impl(PyTypeObject *type, PyObject *value)
+/*[clinic end generated code: output=05af0cfa4805fcae input=3f472ebfd07bbf50]*/
+{
+    uuid_state *state = get_uuid_state_by_cls(type);
+
+    if (validate_int(state, value) < 0) {
+        return NULL;
+    }
+
+    uuidobject *self = make_uuid(type);
+    if (self == NULL) {
+        return NULL;
+    }
+
+    if (from_int(self, value, 0) < 0) {
+        // We validated before creating an instance, so now we don't need to
+        // validate again
+        return NULL;
+    }
+
+    return (PyObject *)self;
+}
+
+static PyGetSetDef Uuid_getset[] = {
+    {"int", (getter)Uuid_get_int, NULL, "UUID as a 128-bit integer", NULL},
```

Comment:
Please remove the casts and change the signatures accordingly. Those functions all lead to undefined behaviours.

Reply:
I don't mind changing, but this is such a standard practice. What's undefined here? All these casts just upcast `PyObject*` to `uuidobject*`?

Reply:
More precisely we want the function signatures to take a PyObject, not a uuidobject if they are object methods (the getters I meant)

Reply:
> Yes. I changed all CPython since last year to avoid those UBs. So I really want to keep this practice.

Wow! Great job, I'll update my code then.